[clang-tidy][abseil-string-find-startswith] Add string_view to default string-like classes

nicovank opened the pull request:
As per title. A small improvement to this check, `string_view` should work out of the box.

I am slowly looking into adding an option to this check to recommend C++20's API `starts_with`, and move it to `modernize`.
Possibly would like to use it as a codemod on our code.

nicovank committed:
[clang-tidy][abseil-string-find-startswith] Add string_view to default string-like classes

nicovank commented on clang-tools-extra/test/clang-tidy/checkers/abseil/string-find-startswith.cpp:
```diff
@@ -22,6 +22,21 @@ struct basic_string {
 typedef basic_string<char> string;
 typedef basic_string<wchar_t> wstring;
 
+template <typename C>
+class basic_string_view {
+public:
+  basic_string_view();
+  basic_string_view(const basic_string_view &);
+  basic_string_view(const C *);
+  ~basic_string_view();
+  int find(basic_string_view s, int pos = 0);
+  int find(const C *s, int pos = 0);
+  int find(const C *s, int pos, int n);
+  int find(char c, int pos = 0);
+  static constexpr size_t npos = -1;
+};
+typedef basic_string_view<char> string_view;
```
https://github.com/llvm/llvm-project/blob/57dd23bc0a2f7b4f7b68162923b3267c1f303de9/clang-tools-extra/test/clang-tidy/checkers/abseil/string-find-str-contains.cpp#L24-L37

nicovank commented:
First time with LLVM GitHub PR workflow. Am I supposed to tag potential reviewers? @gribozavr @hokein @njames93

PiotrZSL commented on clang-tools-extra/test/clang-tidy/checkers/abseil/string-find-startswith.cpp:
```diff
@@ -22,6 +22,21 @@ struct basic_string {
 typedef basic_string<char> string;
 typedef basic_string<wchar_t> wstring;
 
+template <typename C>
```
Actually string and string_view is defined already in header test/clang-tidy/checkers/Inputs/Headers/string.

Just add `-- -isystem %clang_tidy_headers` like tests for readability-string-compare does, and simply include string.

PiotrZSL reviewed (commented):
As for this check:
- use string header
- update check documentation (option default and description)
- make sure that documentation is in sync with class documentation
- add release notes entry about improving check to support std::string_view

PiotrZSL reviewed (changes requested):
documentation, string header, release notes

nicovank commented:
> Copy this check.

Got it. Thanks for the pointer!

> Documentation, string header.

I think I covered everything, had to add a couple of skeletons to that `string` common header.

PiotrZSL commented on clang-tools-extra/clang-tidy/abseil/StringFindStartswithCheck.cpp:
```diff
@@ -19,11 +19,14 @@ using namespace clang::ast_matchers;
 namespace clang::tidy::abseil {
 
+const auto DefaultStringLikeClasses =
+    "::std::basic_string; ::std::basic_string_view";
```
Remove this space
```suggestion
    "::std::basic_string;::std::basic_string_view";
```

PiotrZSL commented on clang-tools-extra/docs/clang-tidy/checks/abseil/string-find-startswith.rst:
```diff
@@ -28,9 +28,9 @@ Options
-   Semicolon-separated list of names of string-like classes. By default only
-   ``std::basic_string`` is considered. The list of methods to considered is
-   fixed.
+   Semicolon-separated list of names of string-like classes. By default both
```
Line 6 in this file should also be changed, something like:
"Checks whether a ``std::string::find()`` or ``std::string::rfind()``  (and corresponding ``std::string_view`` methods)"

Simply should be there some info also about string_view, no need to update examples, just description.

PiotrZSL reviewed (approved):
LGTM, once you correct latest comments & CI will pass, fell free to merge.

nicovank commented:
Addressed feedback.

I do not have commit access, can you please merge after CI is green?
Assuming GitHub will properly populate information, if not you can use `Nicolas van Kempen <nvankempen@meta.com>`.
